Makefile

```diff
@@ -251,9 +251,9 @@ DOCKER_GO = _() { $(SET_X); mkdir -p $(CURDIR)/.go/src/$${3:-dummy} ; mkdir -p $
 
 PARSE_PKGS=$(if $(strip $(EVE_HASH)),EVE_HASH=)$(EVE_HASH) DOCKER_ARCH_TAG=$(DOCKER_ARCH_TAG) ./tools/parse-pkgs.sh
 LINUXKIT=$(BUILDTOOLS_BIN)/linuxkit
-LINUXKIT_VERSION=80c4edd5c54dc05fbeae932440372990fce39bd6
+LINUXKIT_VERSION=ea61ff95ce93871a9bd1ee2fc0a61086b1c51cd4
 LINUXKIT_SOURCE=https://github.com/linuxkit/linuxkit.git
-LINUXKIT_OPTS=--disable-content-trust $(if $(strip $(EVE_HASH)),--hash) $(EVE_HASH) $(if $(strip $(EVE_REL)),--release) $(EVE_REL) $(FORCE_BUILD)
+LINUXKIT_OPTS=$(if $(strip $(EVE_HASH)),--hash) $(EVE_HASH) $(if $(strip $(EVE_REL)),--release) $(EVE_REL) $(FORCE_BUILD)
```

Comment:
Please also add  `--docker --platforms=linux/$(ZARCH)` here and for the rest.
We need images inside the docker as we expect them in our scripts in tools directory. Also we want to use eve image from docker in tests and build workflows.
We want to build only for one arch, not for all linuxkit supports. We can work on cross-compile support in next PRs (we will need to adjust scripts and workflows).

Reply:
I am not sure we do. We only need those for stuff in tools. The pkgs, like pillar, do not need to be in docker at all. Why duplicate the cached data. Or did I misunderstand?

Reply:
It is really hard to predict, we have such line in Makefile: https://github.com/lf-edge/eve/blob/69584aec8eda6e9ad84b6683f718ea25abb9d0e5/Makefile#L367 and we expect existence for most packages in docker as well

Reply:
I am not convinced we do (or that we should). Most (all?) packages from `pkg/` are just used to build other `pkg/` (a few) or runnable OS images. Only some stuff from `tools/` is used for `docker run`.

Reply:
The list of packages we want to see in docker:
From tools:

1. mkconf
2. mkimage-raw-efi
3. mkimage-iso-efi
4. mkrootfs-squash
5. mkrootfs-ext4

From workflows:

1. eve

From Makefile to unpack files:

1. grub
2. u-boot
3. kernel _not sure about new-kernel, cannot find direct usage_
4. ipxe
5. uefi
6. u-boot

From user perspective:

1. edgeview
2. k3s _not sure_

Reply:
@giggsoff does this PR require doing something for the above packages to avoid a regression for the build?

Reply:
We need to load them into docker as well as into linuxkit cache. We can load all packages with adding of `--docker` flag. Or we can implement additional logic to use this flag only for the packages from the list to save space/time.

Reply:
It does. It is on my list to get to it. I assume early next week.

Reply:
@giggsoff some of this makes sense to me, others don't. We may need to open a new issue for it, but let's at least try to do it here.

### tools

The 5 tools you listed: mkconf, mkimage-*, mkroot-*.

As far as I can tell, those are all used in two places:

1. various `tools/*.sh` scripts, which are called by `make` targets to generate eve image outputs. Here they are called via `docker run`.
2. `pkg/eve`, where it copies commands from those same images in order to generate config or rootfs when someone runs the `docker run eve install_*` commands.

For the second, we don't need it in docker cache, as it is just building from one package to the other, i.e. linuxkit cache (which is fixed by this PR).

For the first, we need to be able to run them, so we do need them in docker cache so we can run them. We do it in various parts of the CI pipeline, as well as some `make` commands.

Is that correct?

### eve

For end-users, we do not need it in docker cache. We finish our pipelines by doing `lkt pkg push` for eve, so it ends up in the registry from whence they pull it.

For build pipelines in `assets.yml`, we run `docker run eve` to generate a `rootfs.img` asset which we upload as an asset as part of the release. This actually pulls `eve` from the OCI registry if we are in `lf-edge/eve` repository, and build locally if we are not. If we are not, then there shouldn't be any publishing anyways.

This appears to be the only place we run it. Is there anywhere else? If so, we shouldn't need to load it into docker cache.

### user

edgeview and k3s. 

I did not find edgeview used as `docker run` anywhere. If it is only for clients after pushing to registry, then it doesn't need to be loaded. But I might have missed something. I did find it as included in an eve-os image, but that doesn't require docker loading.

I did not find k3s used as `docker run` either. As far as I can see, k3s is used solely embedded in eve-os, so we should not need to load it.

### Makefile to unpack

You listed:

* grub
* u-boot
* kernel
* ipxe
* uefi

I don't see this run anywhere in `Makefile` via `docker run`, or anywhere in `tools/`. Do these need to be loaded?

Reply:
As far as I can see, we have a few classes of tools here:

* only used inside other packages: these do not need to be loaded into docker
* only used in tools / as part of Makefile: these need to be loaded into docker, but probably should not be in `pkgs/` at all, but elsewhere, to indicate their purpose
* used both in tools/Makefile and as part of other packages: these need to be in `pkgs/`, but also need to be in docker.

The last group is the hardest one, as it is hard to distinguish between those that need to be loaded and those that do not. And, yes, all of this is caused by linuxkit having its own cache, which, in turn, is caused by docker not supporting multi-arch indexes. They will be getting there, but it remains some time away AFAIK. I will be the first to throw out the linuxkit cache when it does; in the meantime, we live with it.

We could simply load _all_ of them into docker, but that will create double space usage for everything, which is an enormous amount of space wasted.

Reply:
> @giggsoff some of this makes sense to me, others don't. We may need to open a new issue for it, but let's at least try to do it here.
> 
> ### tools
> The 5 tools you listed: mkconf, mkimage-_, mkroot-_.
> 
> Is that correct?

Yes, we need them in docker to build parts of the EVE-OS image, so we need them to be inside docker.

> ### eve
> For end-users, we do not need it in docker cache. We finish our pipelines by doing `lkt pkg push` for eve, so it ends up in the registry from whence they pull it.
> 
> For build pipelines in `assets.yml`, we run `docker run eve` to generate a `rootfs.img` asset which we upload as an asset as part of the release. This actually pulls `eve` from the OCI registry if we are in `lf-edge/eve` repository, and build locally if we are not. If we are not, then there shouldn't be any publishing anyways.
> 
> This appears to be the only place we run it. Is there anywhere else? If so, we shouldn't need to load it into docker cache.

We use this image (lfedge/eve) to store artifacts into files [here](https://github.com/lf-edge/eve/blob/7bb501d784bb9697bd0d46c2a79fd431f577d3ac/.github/workflows/build.yml#L66) as well and we need it for local development (we use lfedge/eve image to bootstrap EDEN tests). So it worth to have lfedge/eve in docker cache.

> ### user
> edgeview and k3s.
> 
> I did not find edgeview used as `docker run` anywhere. If it is only for clients after pushing to registry, then it doesn't need to be loaded. But I might have missed something. I did find it as included in an eve-os image, but that doesn't require docker loading.
> 
> I did not find k3s used as `docker run` either. As far as I can see, k3s is used solely embedded in eve-os, so we should not need to load it.

I think you are right here. If user need the image of edgeview or k3s he can build it directly from Dockerfile or use published images. So no need to have in docker cache.

> ### Makefile to unpack
> You listed:
> 
> * grub
> * u-boot
> * kernel
> * ipxe
> * uefi
> 
> I don't see any of these run anywhere in `Makefile` via `docker run`, or anywhere in `tools/`. Do these need to be loaded?

As I mentioned https://github.com/lf-edge/eve/pull/2727#discussion_r927590073 it is not so easy to find the usage. Let me describe. [Here](https://github.com/lf-edge/eve/blob/7bb501d784bb9697bd0d46c2a79fd431f577d3ac/Makefile#L357-L368) we call another rule [DOCKER_UNPACK with docker create inside](https://github.com/lf-edge/eve/blob/7bb501d784bb9697bd0d46c2a79fd431f577d3ac/Makefile#L242) for all the packages from the list. So we need them to be available inside docker.